components/eip681/src/parse.rs

```diff
@@ -0,0 +1,1533 @@
+        //   we can't simply apply them with `number`.or(`address`).
+        // 3. `string` will always match a valid `number`.
+        //
+        // To solve all of these cases we run all three parsers and then return whichever one
+        // consumed the most input.
+        //
+        // In practice this should be fine because values either end with `&` (in the case of
+        // more parameters after it) or with the end of the input.
+
+        let mut results = [number.parse(i), ethereum_address.parse(i), string.parse(i)];
+        results.sort_by_key(|a| {
+            // Sort by the amount of input left over, smallest first
+            a.as_ref().ok().map(|(i, _)| i.len()).unwrap_or(usize::MAX)
+        });
+        results
+            .into_iter()
+            .next()
+            .expect("safe because this was an array of 3")
+    }
+
+    /// Return the value as a [`Number`], if possible.
+    pub fn as_number(&self) -> Option<&Number> {
+        match self {
+            Value::Number(n) => Some(n),
+            _ => None,
+        }
+    }
+}
+
+/// An Ethereum ABI type name.
+///
+/// See [Solidity ABI types](https://docs.soliditylang.org/en/develop/abi-spec.html#types)
+/// for more info.
+///
+/// ## Note
+/// Instead of parsing [`EthereumAbiTypeName`] as a nested syntax tree, we
+/// instead only parse the _name_.
+// TODO(schell): If we do end up wanting a syntax tree representing the type, I
+// have the start of one on the branch `feat/eip-681-tx-req-parser-solidity-types`
+#[derive(Clone, Debug, PartialEq)]
+pub struct EthereumAbiTypeName {
+    name: String,
+}
+
+impl core::fmt::Display for EthereumAbiTypeName {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.name.fmt(f)
+    }
+}
+
+impl EthereumAbiTypeName {
+    /// Parse a `Type`.
+    pub fn parse(i: &str) -> nom::IResult<&str, Self, ParseError<'_>> {
+        fn is_type_char(c: char) -> bool {
+            let chars = &[
+                '[', ']', // Arrays
+                '<', '>', // Type constructors
+                '(', ')', ',', // Tuples
+            ];
+            c.is_alphanum() || chars.contains(&c)
+        }
+        let (i, s) = nom::bytes::complete::take_while1(is_type_char)(i)?;
+        Ok((i, EthereumAbiTypeName { name: s.to_owned() }))
+    }
+}
+
+/// A parameter key.
+///
+/// ```abnf
+/// key = "value" / "gas" / "gasLimit" / "gasPrice" / TYPE
+/// ```
+#[derive(Clone, Debug, PartialEq)]
+pub enum Key {
+    Value,
+    Gas,
+    GasLimit,
+    GasPrice,
+    Type(EthereumAbiTypeName),
+}
+
+impl core::fmt::Display for Key {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Key::Value => f.write_str("value"),
+            Key::Gas => f.write_str("gas"),
+            Key::GasLimit => f.write_str("gasLimit"),
+            Key::GasPrice => f.write_str("gasPrice"),
+            Key::Type(ty) => ty.fmt(f),
+        }
+    }
+}
+
```

Comment:
Then `Key` goes away entirely:
```suggestion
```

Reply:
We still need `Key` for querying parameters, and it is also helpful for parsing - we can parse `Key` first, then match on that for the value, so I've kept it around.